content: draft: Add 'principle' protecting anonymous contribution in SLSA (#1249)
slsa-framework/slsa: docs/spec/draft/principles.md

keep right. There are often over-provisioning, confused deputy problems, or
mistakes. Even if a platform is configured properly at one moment, it might not
stay that way, and humans almost always end up getting in the access control
lists.

Access control is still important, but SLSA goes further to provide defense in depth: it **requires proof in
the form of attestations that the package was built correctly**.

**Benefits**: The attestation removes intermediate platforms from the trust base and ensures that
individuals who are accidentally granted access do not have sufficient permission to tamper with the package.

> ## Support anonymous and pseudonymous contribution

Reviewer: ```suggestion
## Support anonymous and pseudonymous contributions
```

SLSA supports anonymous and pseudonymous 'identities' within the software supply chain.

> While organizations that implement SLSA may chose otherwise, SLSA itself does not require,

Reviewer: ```suggestion
While organizations that implement SLSA may choose otherwise, SLSA itself does not require,
```

or encourage, participants to be mapped to their legal identities.

**Nothing in this specification should be taken to mean that SLSA requires participants to
to reveal their legal identity.**

**Reasoning**: One of SLSA's other principles is to [trust code, not individuals](#trust-code-not-individuals).

> The legal identity of actor is largely not relevant from a supply chain security perspective.  Further,

Reviewer: I support this. Is it possible to set expectations on how identities are meant to be used across SLSA tracks? I'm reading this entire change like so, but perhaps it could be made explicit here?

> **Reasoning**: Some SLSA tracks (e.g., the source track) use identities to attribute supply chain actions to the actor performing them. These identities need to be *strongly authenticated* and *consistent* (i.e., a developer uses the same unique identity for successive supply chain actions) within the context of the supply chain execution and artifact verification. These properties do not require the identities to be mapped to legal identities.

This is a bit hand wavy (can we define strongly authenticated and consistent etc. better? probably), but this might a) clarify why identities don't need to be legal identities, and b) provide grounding for new track authors on how to think about identities.

> The legal identity of actor is largely not relevant from a supply chain security perspective.  Further,

Author: I'm somewhat worried about having the requirements from different tracks seep into the principles (that seems wrong??).

But maybe we can get at it more broadly...

> **Reasoning:** SLSA uses identities for multiple purposes: as a trust anchor for attestations (i.e. who or what is making this claim and do I trust them to do so) or for attributing actions to the actor which took them within the context of the system they're taking them in (e.g. a username, cryptographic signing key, etc...).
> 
> When identities are strongly authenticated and used consistently they can be used for both of these purposes without requiring them to be mapped to legal identities.
> This reflects how identities are often used in open source.  A legal name means much less to projects than the history and behavior of a given handle over time does.

> The legal identity of actor is largely not relevant from a supply chain security perspective.  Further,

Reviewer: This looks good to me!

> The legal identity of actor is largely not relevant from a supply chain security perspective.  Further,

Author: Updated.

stay that way, and humans almost always end up getting in the access control
lists.

Access control is still important, but SLSA goes further to provide defense in depth: it **requires proof in
the form of attestations that the package was built correctly**.

**Benefits**: The attestation removes intermediate platforms from the trust base and ensures that
individuals who are accidentally granted access do not have sufficient permission to tamper with the package.

## Support anonymous and pseudonymous contributions

SLSA supports anonymous and pseudonymous 'identities' within the software supply chain.
While organizations that implement SLSA may choose otherwise, SLSA itself does not require,
or encourage, participants to be mapped to their legal identities.

**Nothing in this specification should be taken to mean that SLSA requires participants to
to reveal their legal identity.**

**Reasoning**: SLSA uses identities for multiple purposes: as a trust anchor for attestations
(i.e. who or what is making this claim and do I trust it to do so) or for attributing actions
to the actor which took them within the context of the system they're taking them in (e.g. a

> username, cryptographic signing key, etc...).

Reviewer 2: ```suggestion
username, cryptographic signing key, etc.).
```
No need to have etc and elipses.

Reviewer 3: > attributing actions
to the actor which took them within the context of the system they're taking them in (e.g. a
username, cryptographic signing key, etc...).

This sentence is a lot!

maybe it would read more simply as: 

```
... or for attributing actions to an actor. Choice of identification technology is left to the platform that provides the action (e.g. username, cryptographic signing key, etc.).
```

Author: > > attributing actions
> > to the actor which took them within the context of the system they're taking them in (e.g. a
> > username, cryptographic signing key, etc...).
> 
> This sentence is a lot!
> 
> maybe it would read more simply as:
> 
> ```
> ... or for attributing actions to an actor. Choice of identification technology is left to the platform that provides the action (e.g. username, cryptographic signing key, etc.).
> ```

I like this

> When identities are strongly authenticated and used consistently they can be used for both of

Reviewer 2: What do you mean by "used consistently"? I feel like trying to deduplicate "used" in one sentence and the first seems like a more likely candidate.

Or how about 
```suggestion
When identities are strongly authenticated and used consistently they can be leveraged for both of
```

> When identities are strongly authenticated and used consistently they can be used for both of

Reviewer 2: Is there a common definition for "strongly authenticated" here? I assume that private signing keys are a form of this strong authentication?

Author: There isn't.  Does there need to be?  This principles page isn't setting any specific requirements.

FWIW "Strongly authenticated" is used here: https://slsa.dev/spec/v1.0/threats#:~:text=Mitigation%3A%20Source%20control%20platform%20strongly%20authenticates%20actor%20identity%2C%20timestamp%2C%20and%20parent%20revisions.

We do talk about strong authentication in the source track (where it is a requirement) but it applies to people not services accounts and here we're also talking about services accounts to some extent.

Reviewer 2: Maybe it doesn't. The previous paragraph seemed more generic and easier to apply to multiple tracks. This paragraph feels more specifically tailored to the source track.

I'll still approve the PR even with this feedback.

Author: Honestly, I cannot manage to navigate GitHub's history to find the version you did like. :/

Do you happen to have a suggestion for what to change here?  Which parts seem specific to the source track?

Reviewer 2: I think the current form looks great. Resolving the thread.

> these purposes without requiring them to be mapped to legal identities.  This reflects how
> identities are often used in open source. A legal name means much less to projects than the
> history and behavior of a given handle over time does.

Reviewer 3: ```suggestion
these purposes without requiring them to be mapped to legal identities.  This reflects how
identities are often used in open source. A legal name means much less to projects than the
history and behavior of a given handle over time does.
```
```suggestion
these purposes without requiring them to be mapped to legal identities.  
This reflects how identities are often used in open source and supports behavior tracking of specific actors over time.
```

Reviewer 3: I think the key idea is that a system must have a way for actors to be uniquely identified and have track records.

Author: I don't love "behavior tracking".  Sounds nefarious and like something folks would object to.

Reviewer 4: I think this is where we might want to leave it up to the consumer. I might be OK with using software from completely anonymous contributors but a bank might not be comfortable. There's nothing they can do about it but not use the software. What SLSA does a good job at is helping all end users be able to make the decisions for themselves whereas previously there might not be a good set of data around this.

Also, I agree that behavior can sound a bit bad but maybe "activity tracking" is better? I think the idea though here is nothing in SLSA itself cares whether or not you're anonymous, pseudonymous, or using a government issued legal identity. However, tracking that is important and both maintainers and end users might care whether or not a project all of a sudden has 10 people with all pseudonymous identities with no known activity history start contributing to a project.

Author: I've added a bit to make it clear organizations can impose additional requirements on identities if they wish.

Since this isn't a requirements doc but just trying to emphasize that SLSA does allow anonymous and pseudonymous contributions I think the ability to track those identities over time is best left to the tracks that need it?

Reviewer 4: Sure, seems reasonable.

mistakes. Even if a platform is configured properly at one moment, it might not
stay that way, and humans almost always end up getting in the access control
lists.

Access control is still important, but SLSA goes further to provide defense in depth: it **requires proof in
the form of attestations that the package was built correctly**.

**Benefits**: The attestation removes intermediate platforms from the trust base and ensures that
individuals who are accidentally granted access do not have sufficient permission to tamper with the package.

## Support anonymous and pseudonymous contributions

SLSA supports anonymous and pseudonymous 'identities' within the software supply chain.
While organizations that implement SLSA may choose otherwise, SLSA itself does not require,
or encourage, participants to be mapped to their legal identities.

**Nothing in this specification should be taken to mean that SLSA requires participants to
to reveal their legal identity.**

**Reasoning**: SLSA uses identities for multiple purposes: as a trust anchor for attestations
(i.e. who or what is making this claim and do I trust it to do so) or for attributing actions

> to an actor. Choice of identification technology is left to the platform that provides the

Reviewer: ```suggestion
to an actor. Choice of identification technology (e.g. username, cryptographic signing key, etc.) is left to the specific instantiation of the SLSA specification.
```

I don't know if this is the right phrasing but I was if we want something other than "platform" as the identity provider.

Reviewer: This looks good to me!
